src/opm/io/eclipse/EGrid.cpp

```diff
@@ -355,4 +355,188 @@ void EGrid::getCellCorners(int globindex, std::array<double,8>& X,
+
+    fileH.open(inputFileName, std::ios::in |  std::ios::binary);
+
+    if (!fileH)
+        throw std::runtime_error("Can not open EclFile: {}" + this->inputFilename);
+
+    std::string arrName(8,' ');
+    eclArrType arrType;
+    int64_t num;
+    int sizeOfElement;
+
+    uint64_t zcorn_pos = 0;
+
+    while (!isEOF(&fileH)) {
+
+        readBinaryHeader(fileH,arrName,num, arrType, sizeOfElement);
+
+        if (arrName == "ZCORN   "){
+            zcorn_pos = fileH.tellg();
+            break;
+        }
+
+        uint64_t sizeOfNextArray = sizeOnDiskBinary(num, arrType, sizeOfElement);
+        fileH.seekg(static_cast<std::streamoff>(sizeOfNextArray), std::ios_base::cur);
+    }
+
+    int elements_pr_block = Opm::EclIO::MaxBlockSizeReal / Opm::EclIO::sizeOfReal;
+    int num_blocks_start = zcorn_offset / elements_pr_block;
+
+    // adding size of zcorn real data before to ignored
+    uint64_t start_pos = zcorn_pos + Opm::EclIO::sizeOfReal * zcorn_offset;
+
+    // adding size of blocks (head and tail flags)
+    start_pos = start_pos + (1 + num_blocks_start * 2) * Opm::EclIO::sizeOfInte;
+
+    fileH.seekg(start_pos, std::ios_base::beg);
+
+    float value;
+    int zcorn_to = zcorn_offset + nodes_pr_surf;
+
+    for (int ind = zcorn_offset; ind < zcorn_to; ind++) {
+
+        if ((ind > 0) && ((ind % elements_pr_block) == 0))
+            fileH.seekg(static_cast<std::streamoff>(2*Opm::EclIO::sizeOfInte), std::ios_base::cur);
+
+        fileH.read(reinterpret_cast<char*>(&value), Opm::EclIO::sizeOfReal);
+        value = Opm::EclIO::flipEndianFloat(value);
+
+        zcorn_layer.push_back(value);
+    }
+
+    fileH.close();
+
+    return zcorn_layer;
+}
+
+void EGrid::getCellCorners(const std::array<int, 3>& ijk, const std::vector<float>& zcorn_layer,
+                           std::array<double,4>& X, std::array<double,4>& Y, std::array<double,4>& Z)
+{
+    std::vector<int> zind;
+    std::vector<int> pind;
+
+   // calculate indices for grid pillars in COORD arrray
+    pind.push_back(ijk[1]*(nijk[0]+1)*6 + ijk[0]*6);
+    pind.push_back(pind[0] + 6);
+    pind.push_back(pind[0] + (nijk[0]+1)*6);
+    pind.push_back(pind[2] + 6);
+
+    // get depths from zcorn array in ZCORN array
+    zind.push_back(ijk[2]*nijk[0]*nijk[1]*8 + ijk[1]*nijk[0]*4 + ijk[0]*2);
+    zind.push_back(zind[0] + 1);
+    zind.push_back(zind[0] + nijk[0]*2);
+    zind.push_back(zind[2] + 1);
+
+    for (int n = 0; n< 4; n++)
+        Z[n] = zcorn_layer[zind[n]];
+
+    for (int  n = 0; n < 4; n++) {
+        double xt;
+        double yt;
+        double xb;
+        double yb;
+
+        double zt = coord_array[pind[n] + 2];
+        double zb = coord_array[pind[n] + 5];
+
+        xt = coord_array[pind[n]];
+        yt = coord_array[pind[n] + 1];
+        xb = coord_array[pind[n] + 3];
+        yb = coord_array[pind[n] + 4];
+
+        X[n] = xt + (xb-xt) / (zt-zb) * (zt - Z[n]);
+        Y[n] = yt+(yb-yt)/(zt-zb)*(zt-Z[n]);
```

Comment:
Do we need to guard against `zt==zb` here?  It's an unlikely but nevertheless possible case if I recall correctly.  In particular, some modelling packages like to represent vertical pillars as
```
(xt == xb) && (yt == yb) && (zt == zb)
```